### PR TITLE
replace the leak check with universes, take 2

### PR DESCRIPTION
**Pull request by nikomatsakis**
This PR is an attempt to revive the "universe-based region check", which is an important step towards lazy normalization. Unlike before, we also modify the definition of `'empty` so that it is indexed by a universe. This sidesteps some of the surprising effects we saw before -- at the core, we no longer think that `exists<'a> { forall<'b> { 'b: 'a } }` is solveable. The new region lattice looks like this:

```
static ----------+-----...------+       (greatest)
|                |              |
early-bound and  |              |
free regions     |              |
|                |              |
scope regions    |              |
|                |              |
empty(root)   placeholder(U1)   |
|            /                  |
|           /         placeholder(Un)
empty(U1) --         /
|                   /
...                /
|                 /
empty(Un) --------                      (smallest)
```
This PR has three effects:

* It changes a fair number of error messages, I think for the better.
* It fixes a number of bugs. The old algorithm was too conservative and caused us to reject legal subtypings.
* It also causes two regressions (things that used to compile, but now do not).
    * `coherence-subtyping.rs` gets an additional error. This is expected.
    * `issue-57639.rs` regresses as before, for the reasons covered in #57639.

Both of the regressions stem from the same underlying property: without the leak check, the instantaneous "subtype" check is not able to tell whether higher-ranked subtyping will succeed or not. In both cases, we might be able to fix the problem by doing a 'leak-check like change' at some later point (e.g., as part of coherence).

This is a draft PR because:

* I didn't finish ripping out the leak-check completely.
* We might want to consider a crater run before landing this.
* We might want some kind of design meeting to cover the overall strategy.
* I just remembered I never finished 100% integrating this into the canonicalization code.
* I should also review what happens in NLL region checking -- it probably still has a notion of bottom (empty set).

r? @matthewjasper

### CODE REVIEW

File: src/librustc/infer/lexical_region_resolve/mod.rs [hunk 1]
```diff
@@ -83,6 +83,16 @@ pub enum RegionResolutionError<'tcx> {
         Region<'tcx>,
     ),
 
+    /// Indicates a `'b: 'a` constraint where `'a` is in a universe that
+    /// cannot name the placeholder `'b`
```
Comment (Centril) [inline src/librustc/infer/lexical_region_resolve/mod.rs]:
```suggestion
    /// cannot name the placeholder `'b`.
```

File: src/librustc/infer/lexical_region_resolve/mod.rs [hunk 2]
```diff
@@ -402,14 +419,40 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
...
+        let sub_free_regions = |r1, r2| self.region_rels.free_regions.sub_free_regions(tcx, r1, r2);
+
+        // Check for the case where we know that `'b: 'static` -- in that case,
+        // `a <= b` for all `a`.
+        let b_free_or_static = self.region_rels.free_regions.is_free_or_static(b);
+        if b_free_or_static && sub_free_regions(tcx.lifetimes.re_static, b) {
+            return true;
+        }
+
+        // If both a and b are free, consult the declared
```
Comment (Centril) [inline src/librustc/infer/lexical_region_resolve/mod.rs]:
```suggestion
        // If both `a` and `b` are free, consult the declared
```

File: src/librustc/infer/lexical_region_resolve/mod.rs [hunk 3]
```diff
@@ -402,14 +419,40 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
...
+        // `lub` relationship defined below, since sometimes the "lub"
+        // is actually the `postdom_upper_bound` (see
+        // `TransitiveRelation` for more details).
+        let a_free_or_static = self.region_rels.free_regions.is_free_or_static(a);
+        if a_free_or_static && b_free_or_static {
+            return sub_free_regions(a, b);
+        }
+
+        // For other cases, leverage the LUB code to find the LUB and
+        // check if it is equal to b.
```
Comment (Centril) [inline src/librustc/infer/lexical_region_resolve/mod.rs]:
```suggestion
        // check if it is equal to `b`.
```

File: src/librustc/infer/lexical_region_resolve/mod.rs [hunk 4]
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
                 );
             }
 
+            (&ReStatic, _) | (_, &ReStatic) => {
+                // nothing lives longer than static
```
Comment (Centril) [inline src/librustc/infer/lexical_region_resolve/mod.rs]:
```suggestion
                // Nothing lives longer than `static`.
```

File: src/librustc/infer/lexical_region_resolve/mod.rs [hunk 5]
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
...
+                tcx.lifetimes.re_static
+            }
+
+            (&ReEmpty(_), r @ ReEarlyBound(_)) | (r @ ReEarlyBound(_), &ReEmpty(_))
+            | (&ReEmpty(_), r @ ReFree(_)) | (r @ ReFree(_), &ReEmpty(_))
+            | (&ReEmpty(_), r @ ReScope(_)) | (r @ ReScope(_), &ReEmpty(_)) => {
+                // all empty regions are less than early-bound, free,
```
Comment (Centril) [inline src/librustc/infer/lexical_region_resolve/mod.rs]:
```suggestion
                // All empty regions are less than early-bound, free,
```

File: src/librustc/infer/lexical_region_resolve/mod.rs [hunk 6]
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
...
+                // and scope regions
```
Comment (Centril) [inline src/librustc/infer/lexical_region_resolve/mod.rs]:
```suggestion
                // and scope regions.
```

File: src/librustc/infer/lexical_region_resolve/mod.rs [hunk 7]
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
...
+                r
+            }
+
+            (&ReEmpty(a_ui), &ReEmpty(b_ui)) => {
+                // empty regions are ordered according to the universe
```
Comment (Centril) [inline src/librustc/infer/lexical_region_resolve/mod.rs]:
```suggestion
                // Empty regions are ordered according to the universe
```

File: src/librustc/infer/lexical_region_resolve/mod.rs [hunk 8]
```diff
@@ -437,6 +472,38 @@ impl<'cx, 'tcx> LexicalResolver<'cx, 'tcx> {
...
+                // they are associated with
```
Comment (Centril) [inline src/librustc/infer/lexical_region_resolve/mod.rs]:
```suggestion
                // they are associated with.
```

File: src/librustc/infer/outlives/free_region_map.rs [hunk 9]
```diff
@@ -23,11 +23,65 @@ impl<'tcx> FreeRegionMap<'tcx> {
...
+            // `'a <= 'static` is always true, and not stored in the
+            // relation explicitly, so check if `'b` is `'static` (or
+            // equivalent to it)
+            true
+        } else {
+            self.check_relation(r_a, r_b)
+        }
+    }
+
+    /// Check whether `r_a <= r_b` is found in the relation
```
Comment (Centril) [inline src/librustc/infer/outlives/free_region_map.rs]:
```suggestion
    /// Check whether `r_a <= r_b` is found in the relation.
```

File: src/librustc/infer/outlives/verify.rs [hunk 10]
```diff
@@ -64,7 +64,18 @@ impl<'cx, 'tcx> VerifyBoundCx<'cx, 'tcx> {
...
-        VerifyBound::AnyBound(param_bounds.map(|r| VerifyBound::OutlivedBy(r)).collect())
+        let any_bounds: Vec<_> = param_bounds.map(|r| VerifyBound::OutlivedBy(r)).collect();
+
+        if any_bounds.is_empty() {
+            // We know that all types `T` outlive `'empty`, so if we
+            // can find no other bound, then check that the region
+            // being tested is `'empty`.
+            VerifyBound::IsEmpty
+        } else {
+            // If we can find any other bound R such that `T: R`, then
```
Comment (Centril) [inline src/librustc/infer/outlives/verify.rs]:
```suggestion
            // If we can find any other bound `R` such that `T: R`, then
```

File: src/librustc/ty/context.rs [hunk 11]
```diff
@@ -181,8 +181,13 @@ pub struct CommonTypes<'tcx> {
 }
 
 pub struct CommonLifetimes<'tcx> {
-    pub re_empty: Region<'tcx>,
+    /// ReEmpty in the root universe
```
Comment (Centril) [inline src/librustc/ty/context.rs]:
```suggestion
    /// `ReEmpty` in the root universe.
```

File: src/librustc/ty/context.rs [hunk 12]
```diff
@@ -181,8 +181,13 @@ pub struct CommonTypes<'tcx> {
...
+    pub re_root_empty: Region<'tcx>,
+
+    /// ReStatic
```
Comment (Centril) [inline src/librustc/ty/context.rs]:
```suggestion
    /// `ReStatic`
```

File: src/librustc/ty/sty.rs [hunk 13]
```diff
@@ -1301,14 +1357,13 @@ pub enum RegionKind {
...
-    /// specially; at least right now, we do not generate instances of
-    /// it during the GLB computations, but rather
-    /// generate an error instead. This is to improve error messages.
-    /// The only way to get an instance of ReEmpty is to have a region
-    /// variable with no constraints.
-    ReEmpty,
+    /// Empty lifetime is for data that is never accessed.  We tag the
+    /// empty lifetime with a universe -- the idea is that we don't
+    /// want `exists<'a> { forall<'b> { 'b: 'a } }` to be satisfiable.
+    /// Therefore, the `'empty` in a universe U is less than all
```
Comment (Centril) [inline src/librustc/ty/sty.rs]:
```suggestion
    /// Therefore, the `'empty` in a universe `U` is less than all
```

File: src/librustc/ty/sty.rs [hunk 14]
```diff
@@ -1301,14 +1357,13 @@ pub enum RegionKind {
...
+    /// regions visible from U, but not less than regions not visible
```
Comment (Centril) [inline src/librustc/ty/sty.rs]:
```suggestion
    /// regions visible from `U`, but not less than regions not visible
```

File: src/librustc/ty/sty.rs [hunk 15]
```diff
@@ -1301,14 +1357,13 @@ pub enum RegionKind {
...
+    /// from U.
```
Comment (Centril) [inline src/librustc/ty/sty.rs]:
```suggestion
    /// from `U`.
```

**Review by Centril — COMMENTED** [review]
Just a series of nits that can be batch-applied.

**Comment by lqd** [comment]
When this PR is complete (particularly the "I didn't finish ripping out the leak-check completely." bullet point), it will also fix #59490.

**Comment by matthewjasper** [comment]
> * I should also review what happens in NLL region checking -- it probably still has a notion of bottom (empty set).

It definitely considers regions that are not live anywhere and don't outlive any universal regions as `'empty`:

https://github.com/rust-lang/rust/blob/9c5a5c471aa9229b82bafa8ec18e55722d9ba8c3/src/librustc_mir/borrow_check/nll/region_infer/mod.rs#L1524-L1533

**Comment by nikomatsakis** [comment]
@matthewjasper yeah, [see also this comment](https://github.com/rust-lang/rust/issues/57642#issuecomment-540012551)

**Comment by nikomatsakis** [comment]
It occurs to me that updating NLL to match is not (presently) a strict requirement, but it would be something we have to do before we disable the AST-based regionck.

**Comment by matthewjasper** [comment]
Can you bless the NLL compare mode output and check how this handles #57936 as well?

**Comment by nikomatsakis** [comment]
@matthewjasper I will bless compare-mode output, yes. As far as #57936, I get two errors: one in the `indirect` function and one in the `direct` function. The `indirect` error appears to be new? It is as follows:

```
error[E0308]: mismatched types
  --> /home/nmatsakis/tmp/issue-57936.rs:19:5
   |
19 |     indirect::<fn(&())>(); // OK
   |     ^^^^^^^^^^^^^^^^^^^ one type is more general than the other
   |
   = note: expected type `X`
              found type `X`
```

This error is very bad -- I think we have an open issue on it, actually? I should fix that. The problem here is that the real cause has to do with the full trait ref, but for some reason we're not printing all the details.

**Comment by nikomatsakis** [comment]
(I had not noticed #57936 at the time, or else I forgot about it, I'm not sure. :)

**Comment by nikomatsakis** [comment]
Commit just pushed. Message:

--bless --compare-mode=nll -- but note that we get two errors

```
failures:
    [ui (nll)] ui/hrtb/hrtb-exists-forall-trait-contravariant.rs
    [ui (nll)] ui/hrtb/issue-46989.rs
```

Both are related to the NLL region checker not handling the `empty(ui)` correctly.

**Comment by lqd** [comment]
> ```
> ...
>    = note: expected type `X`
>               found type `X`
> ```
> 
> This error is very bad -- I think we have an open issue on it, actually? 

We used to have #57362 for example, but it was fixed.

At the time we mentioned there might still be some cases missing from `try_report_placeholder_conflict`'s [big match](https://github.com/rust-lang/rust/blob/master/src/librustc/infer/error_reporting/nice_region_error/placeholder_error.rs#L20-L162) but didn't have cases exercising these possibly missing arms, maybe that is such a case ?

**Comment by nikomatsakis** [comment]
@lqd 

> We used to have #57362 for example, but it was fixed.

OK, well, I guess I broke it again. Thanks for the cite.

File: src/test/ui/issues/issue-57362-2.stderr [hunk 16]
```diff
@@ -1,13 +1,27 @@
...
    |
 LL |     let x = <fn (&())>::make_g();
-   |                         ^^^^^^ function or associated item not found in `for<'r> fn(&'r ())`
+   |             ^^^^^^^^^^^^^^^^^^ one type is more general than the other
    |
-   = help: items from traits can only be used if the trait is implemented and in scope
-   = note: the following trait defines an item `make_g`, perhaps you need to implement it:
-           candidate #1: `X`
+   = note: expected type `X`
+              found type `X`
```
[the remark this answers is not in the document]
Comment (nikomatsakis) [reply src/test/ui/issues/issue-57362-2.stderr]: Yeah, I think before I land this I have to figure out what's going on there

**Comment by lqd** [comment]
I looked a bit at the region errors managing to avoid the nice_region_errors mechanism, and left a couple notes in [this Zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/leak.20check.20to.20universes.20take.202.20.2365232).

**Comment by JohnCSimon** [comment]
Ping from triage.
@nikomatsakis can you please address the merge conflicts? 
Thank you!

**Comment by matthewjasper** [comment]
> * I just remembered I never finished 100% integrating this into the canonicalization code.

If this is referring to this, I guess this can be done in a follow up issue
https://github.com/rust-lang/rust/blob/94fe66d3df05a5f6329bae64e02dfe5ba09eb71a/src/librustc/infer/canonical/canonicalizer.rs#L202-L204

I think that this should be good to go with follow up issues raised and a crater run.

**Comment by JohnCSimon** [comment]
Ping from triage:
@nikomatsakis Can you please address the failing build? 
cc: @matthewjasper @Centril 
Thanks.

**Comment by nikomatsakis** [comment]
@matthewjasper I have rebased this PR and I have converted it to issue warnings. I made them deny-by-default to start. We should probably do a crater run. Deny-by-default seemed reasonable since the last time we made this transition, we went straight to a hard error (see history in https://github.com/rust-lang/rust/issues/56105), owing to the clean crater run. That change was later backed out for independent reasons, but it didn't seem like there were a lot of people complaining in the interim. (At least based on the comments from the tracking issue.)

**Comment by nikomatsakis** [comment]
I am however still running x.py test --bless...

**Comment by nikomatsakis** [comment]
I probably won't be at keyboard when try build completes, but if somebody starts crater I'd be much obliged :)

**Comment by nikomatsakis** [comment]
@craterbot check

**Comment by Centril** [comment]
Looks like the last regression is spurious, so only 2, which is pretty good. The only notable regression is `wasm-bindgen-0.2.58`. I think if we can fix upstream on nightly we should be able to land this without a warning period. cc @ashleygwilliams @fitzgen

**Comment by nikomatsakis** [comment]
Based on these crater results, I think we should land this "as is" but with the FCW set to warn. We can then begin work on follow-up PRs that cleanup the support in NLL and which turn this into a hard error (or perhaps a deny-by-default lint).

Skimming the history, I saw these work items:

* extend support in NLL
* extend support in canonicalization, maybe (see [this comment](https://github.com/rust-lang/rust/pull/65232#issuecomment-553001315))

@matthewjasper also had a number of review comments that are now outdated. I'll have to rebase and see if they still apply.

@matthewjasper, do you want to review this "for real" post rebase? I forget if you r+'d in some form, I didn't notice any comment to that effect earlier.

**Comment by matthewjasper** [comment]
I'll try to go through this one last time once you've rebased

**Comment by nikomatsakis** [comment]
Also, to clarify for @ashleygwilliams and @fitzgen:

* Landing this PR will trigger warnings in wasm-bindgen
* I want to investigate exactly what's going on there
* The old code was accepting things that should never have been accepted
* This code is currently the **most conservative** it can be (i.e., it warns for all cases that may ultimately want to change)
* There is room to make it **less conservative** but that will require a bit more work and it would be useful to see what kinds of things start to get warnings
* So TL;DR I don't think any action is needed immediately for wasm-bindgen but obviously we should resolve that before we make this a hard error

**Comment by fitzgen** [comment]
I'm not really sure what this PR is changing in a user-visible fashion -- is there some sort of summary I can look at somewhere to help understand what needs to change in `wasm-bindgen`?

The linked code that breaks in wasm-bindgen is our implementation of turning Rust closures into JS closures, FWIW. We have a `WasmClosure` trait that describes how to invoke the closure from across a raw wasm abi boundary. We implement it for a bunch of `dyn Fn(..)` and `dyn mut FnMut(...)` things inside a macro because we don't have a way to implement it once for all different function arities so things get pretty repetitive.

**Comment by fitzgen** [comment]
+cc @alexcrichton

**Comment by matthewjasper** [comment]
@fitzgen 
The issue that the following impls (and similar impls for `dyn FnMut` and of 4 or so other traits):

```rust
// Explicitly writing the bound lifetime.
impl<'a, 'b, A, R> IntoWasmAbi for &'a (dyn for<'x> Fn(&'x A) -> R + 'b)
where
    A: RefFromWasmAbi,
    R: ReturnWasmAbi
{ /* ... */ }
// From a macro:
impl<'a, 'b, A, R> IntoWasmAbi for &'a (dyn Fn(A) -> R + 'b)
    where A: FromWasmAbi,
        R: ReturnWasmAbi
{ /* ... */ }
```
Currently these are considered not to overlap because we don't consider it possible for `A` in the second impl to equal `&'x A` in the first.

The simplest way to work around this is to replace the first impl with separate impls for all of the actual types that implement `RefFromWasmAbi`, which would allow the compiler to use its knowledge of which types implement `FromWasmAbi` to ensure that there's no overlap.

@nikomatsakis Can you check that NLL compare mode works?

**Comment by fitzgen** [comment]
> The simplest way to work around this is to replace the first impl with separate impls for all of the actual types that implement `RefFromWasmAbi`, which would allow the compiler to use its knowledge of which types implement `FromWasmAbi` to ensure that there's no overlap.

Unfortunately, every type that downstream users expose to JS via

```rust
#[wasm_bindgen]
pub struct MyTypeExposedToJs {
    // ...
}
```

implements `RefFromWasmAbi` so listing them all is impossible for us.

Is there another work around?

**Comment by nikomatsakis** [comment]
> Currently these are considered not to overlap because we don't consider it possible for `A` in the second impl to equal `&'x A` in the first.

The question, I guess, is whether we *ought* to be warning in cases like these. 

Two considerations:

* We don't have variance-based subtyping for `dyn` types, so really I think it's true that there is no type that could be used for `A` to make it equal to `for<'x> &'x A`.
* For `fn` types, which do have variance for their argument types, the story could be different: if we have a `'empty` region (which I think is not fully settled?), then there would be overlap, but otherwise not.

Does this sound right? (This is partly what I meant by, we might ultimately tune the code to accept more, particularly if there is extant code that we really feel we have to continue accepting, and there is no easy workaround, which might be true here.)

Leave aside for a moment the way we would implement this distinction. But basically I think the question is something like this: equating the argument types yields some set of region constraints RC. Right now in the PR, we assume they are always satisfiable, but we could try to provide they are not satisfiable using *some* decision procedure.

**Comment by alexcrichton** [comment]
Reading over this I don't know a ton about all the liftimes/checks/etc, but it sounds like this will add a warning to the compilation of the wasm-bindgen crate and that's pretty much it. If that's the case I think it's totally fine to go ahead and land this.

Also from reading this I don't think it's going to be easy/clear how we're going to fix the warning in wasm-bindgen (and let rustc transition to a hard error at some point), but it may be best to start a dedicated thread for that elsewhere?

**Comment by nikomatsakis** [comment]
I agree that this doesn't block us from landing, but on the other hand I do think we have to resolve the question of how to migrate wasm bindgen. Currently I suspect the answer is that we'll want to alter the check and accept the code (and I believe we can, though I have to think a bit about how best to implement it).

**Comment by nikomatsakis** [comment]
@matthewjasper `x.py test` and `x.py test --compare-mode=nll` both work for me :)

**Comment by mark-i-m** [comment]
@matthewjasper Perhaps retry?

**Comment by lqd** [comment]
> The install of msys2 was NOT successful.

@bors retry

**Comment by jonas-schievink** [comment]
This [regressed](https://perf.rust-lang.org/compare.html?start=a29424a2265411dda7d7446516ac5fd7499e2b55&end=8498c5f5b02dbb4ed58a1eb4901b0b733342c35f&stat=instructions:u) the `packed-simd` benchmark by ~5%, presumably due to the extra call to `traits::overlapping_impls`.

**Comment by jonas-schievink** [comment]
I have opened a fix at https://github.com/rust-lang/rust/pull/69044
